bwt/bwt.go

```diff
@@ -0,0 +1,424 @@
+some kind of point of reference of where we are. We can do this by storing
+the position of each original character for each of the corresponding
+characters in the F column. With our banana example:
+
+F:   $0 a0 a1 a2 b0 n0 n1
+SA: [6  5  3  1  0  4  2]
+
+If we take our count example for the pattern "ana" above, you'll remember
+that our final search range was [1, 2). If we look up 1 in the SA, we'll
+fund that there is only one offset at position 3 in the original sequence
+"banana"
+
+## Notes on Performance
+
+The explanation above leads to a very naive implementation. For example,
+having the full SA would take way more memory than the BWT itself. Assuming
+int64, that would 8 times the amount of memory of the BWT in its plain text
+representation! In the implementation below, we may instead sample the SA
+and do additional look ups as needed to find the offsets we need.
+
+Similarly, storing the F and L column as plain text has just doubled the
+amount of memory from the original sequence... BWT is used for text
+compression, not expansion! That's why in the below implementation, you
+will see other data structures to actually compress the amount of memory
+needed. You will also notice that we can make huge improvements by
+compressing sequences like with the F column.
+
+Instead of:
+
+F: $0 a0 a1 a2 b0 n0 n1
+
+Since F is lexicographically sorted, we can have:
+
+F: {$: [0, 1)}, {a: [1, 4)}, {b: [4, 5)} {n: [5, 7)}
+
+Although these performance enhancements may look different from what is
+described above, it is still just an FL mapping at the end of the day- just
+with more steps.
+
+
+NOTE: The above is just to explain what is happening at a high level. Please
+reference the implementation below to see how the BWT is actually currently
+working
+*/
+
+const nullChar = "$"
+
+// BWT Burrow Wheeler Transform
+// Compresses and Indexes a given sequence so that it can be
+// be used for search, alignment, and text extraction. This is
+// useful for sequences so large that it would be beneficial
+// to reduce its memory footprint while also maintaining a way
+// to analyze and work with the sequence.
+type BWT struct {
+	// firstColumnSkipList is the first column of the BWT. It is
+	// represented as a list of skipEntries because the first column of
+	// the BWT is always lexicographically ordered. This saves time and memory.
+	firstColumnSkipList []skipEntry
+	// Column last column of the BWT- the actual textual representation
+	// of the BWT.
+	lastCoulmn waveletTree
+	// suffixArray an array that allows us to map a position in the first
+	// column to a position in the original sequence. This is needed to be
+	// able to extract text from the BWT.
+	suffixArray []int
+}
+
+// Count represents the number of times the provided pattern
+// shows up in the original sequence.
+func (bwt BWT) Count(pattern string) (count int, err error) {
+	defer bwtRecovery("Count", &err)
```

Comment:
Spoke a bit about it [here](https://github.com/bebop/poly/pull/411#discussion_r1421267162)

There are some panics that happen in the non public facing parts of the BWT that didn't seem appropriate to bubble up like trying to access some index that it out of bounds on a bitvector etc. 

That Go blog post mentioned what the encoding/json package does internally. It seems important to give users the choice on how to handle the error- especially since they could be doing something very important or time consuming like Tim mentioned.